src/ligero/utils.rs

```diff
@@ -116,32 +113,22 @@ impl<F: Field> Matrix<F> {
 // TODO batch for all rows? possibly minimal savings due to not having to create fft_domain every time
 // here we restrict F to PrimeField
 pub(crate) fn reed_solomon<F: FftField>(
+    // msg, of length m, is interpreted as a vector of coefficients of a polynomial of degree m - 1
     msg: &[F],
     rho_inverse: usize,
-    fft_domain: GeneralEvaluationDomain<F>,
-    domain_iter: &mut GeneralElements<F>,
 ) -> Vec<F> {
-    // TODO is this check worth it?
-    // rho_inverse = 0 should never happen; rho_inverse = 1 means no expansion
-    if rho_inverse <= 1 {
-        return msg.to_vec();
-    }
-
     let m = msg.len();
 
-    let pol: DensePolynomial<F> = DensePolynomial::from_coefficients_slice(&fft_domain.ifft(msg));
-
-    let mut encoding = msg.to_vec();
+    let domain = GeneralEvaluationDomain::<F>::new(m).unwrap();
+    let poly_coeffs = domain.ifft(msg).to_vec();
+    let mut out = msg.to_vec();
+    assert!(out.len() == m);
 
-    domain_iter.nth(m - 1);
-
-    for _ in 0..(rho_inverse - 1) * m {
-        // TODO make sure the domain has enough elements in the caller or check here
-        let zeta = domain_iter.next().unwrap();
-        encoding.push(pol.evaluate(&zeta));
-    }
-
-    encoding
+    // TODO if rho_inv == 2, then with a systematic encoding the second half will be the same as first - is that an issue?
```

Comment:
Why is this the case? Afaik systematic just means that the first entries of the encoded message coincide with the entries of the message itself.

Reply:
Yes, but since we're treating the first entries as the evaluations of some polynomial at `omega_0`, ..., `omega_{m-1}`, the next domain we create is of exactly the same size, i.e. its elements will be `omega_0`, ..., `omega_{m-1}`.
Whereas if `rho_inv != 2`, we get a different omega, so the evaluations at these omegas will be different.
*where omega is a root of unity